docs_nnx/guides/flax_gspmd.md

```diff
@@ -0,0 +1,378 @@
+
+2. Annotate each axis with a name using the `axis_names` parameter in `jax.sharding.Mesh`. A typical way to annotate axis names is `axis_name=('data', 'model')`, where:
+  * `'data'`: the mesh dimension used for data-parallel sharding of the batch dimension of inputs and activations.
+  * `'model'`: the mesh dimension used for sharding parameters of the model across devices.
+
+```{code-cell} ipython3
+# Create a mesh of two dimensions and annotate each axis with a name.
+mesh = Mesh(devices=np.array(jax.devices()).reshape(2, 4),
+            axis_names=('data', 'model'))
+print(mesh)
+```
+
+## Define a model with specified sharding
+
+Create an example layer called `DotReluDot`. This layer carries out two dot product multiplication upon the input `x`, and uses the `jax.nn.relu` (ReLU) activation function in-between.
+
+To annotate a model variable with their ideal sharding, you can use [`flax.nnx.with_partitioning`](https://flax-nnx.readthedocs.io/en/latest/api_reference/flax.nnx/spmd.html#flax.nnx.with_partitioning) to wrap over its initializer function. Essentially, this calls [`flax.nnx.with_metadata`](https://flax-nnx.readthedocs.io/en/latest/api_reference/flax.nnx/variables.html#flax.nnx.with_metadata) and adds a `.sharding` attribute field to the corresponding `nnx.Variable`.
+
+Note that this annotation will be [preserved and adjusted accordingly across lifted transformations](https://flax-nnx.readthedocs.io/en/latest/guides/transforms.html#axes-metadata). This means if you use sharding annotations along with any transform that modifies axes (like `nnx.vmap`, `nnx.scan`), you need to provide sharding of that additional axis via the `transform_metadata` arg. Check out the [NNX transforms guide](https://flax-nnx.readthedocs.io/en/latest/guides/transforms.html) to learn more.
+
+```{code-cell} ipython3
+class DotReluDot(nnx.Module):
+  def __init__(self, depth: int, rngs: nnx.Rngs):
+    init_fn = nnx.initializers.lecun_normal()
+
+    # Initialize a sublayer `self.dot1` and annotate its kernel with
+    # sharding (None, 'model')
+    self.dot1 = nnx.Linear(
+      depth, depth,
+      kernel_init=nnx.with_partitioning(init_fn, (None, 'model')),
+      use_bias=False,  # or use `bias_init` to give it annotation too
+      rngs=rngs)
+
+    # Initialize a weight param `w2` and annotate with sharding ('model', None)
+    self.w2 = nnx.Param(
+      nnx.with_partitioning(init_fn, ('model', None))(
+        rngs.params(), (depth, depth))   # RNG key and shape for W2 creation
+    )
+
+  def __call__(self, x: jax.Array):
+    y = self.dot1(x)
+    y = jax.nn.relu(y)
+    # In data parallelism, input / intermediate value's first dimension (batch)
+    # will be sharded on `data` axis
+    y = jax.lax.with_sharding_constraint(y, PartitionSpec('data', 'model'))
+    z = jnp.dot(y, self.w2.value)
```

Comment:
Variables can be used as JAX arrays thanks to the  `__jax_aray__` protocol.
```suggestion
    z = jnp.dot(y, self.w2)
```

Reply:
Done.